specs/ingress/tests.go

```diff
@@ -58,50 +59,219 @@ func getExternalIP(svc, ns string) (string, error) {
+	default:
+		_, err := h.Kubectl("", "apply", "-f", controllerYAML)
+		gomega.Expect(err).Should(gomega.BeNil(), fmt.Sprintf("%s: %s", installFailMessage, utils.Err(err)))
+
+		err = h.CheckPods(controllerNs, controllerDeployName, 1)
+		gomega.Expect(err).Should(gomega.BeNil(), fmt.Sprintf("failed to verify %s controller pods: %s", ingressName, utils.Err(err)))
+
+		out, err := h.Kubectl("", "get", "-n", controllerNs, "deploy", controllerDeployName, "-o", "yaml")
+		gomega.Expect(err).Should(gomega.BeNil(), fmt.Sprintf("failed to get YAML manifest for deploy/%s: %s", controllerDeployName, utils.Err(err)))
+
+		out, stderr, err := h.PipeToLinkerdRun(out, "inject", "-")
+		gomega.Expect(err).Should(gomega.BeNil(), fmt.Sprintf("failed to inject controller: %s", stderr))
+
+		_, err = h.KubectlApply(out, controllerNs)
+		gomega.Expect(err).Should(gomega.BeNil(), fmt.Sprintf("failed to apply injected manifests: %s", utils.Err(err)))
+
+		err = h.CheckPods(controllerNs, controllerDeployName, 1)
+		gomega.Expect(err).Should(gomega.BeNil(), fmt.Sprintf("failed to verify %s controller pods: %s", ingressName, utils.Err(err)))
+
+		err = utils.CheckProxyContainer(controllerDeployName, controllerNs)
+		gomega.Expect(err).Should(gomega.BeNil(), fmt.Sprintf("controller deployment does not contain proxy sidecar: %s", utils.Err(err)))
+	}
+}
+
+func testControllerUninstall(ingressName, controllerYAML, controllerNs string) {
+	h, _ := utils.GetHelperAndConfig()
+
+	ginkgo.By(fmt.Sprintf("Uninstalling %s ingress controller", ingressName))
+	failMessage := "could not delete ingress controller"
+
+	switch ingressName {
+	case gloo:
+		_, err := utils.GlooctlRun("uninstall", "gateway")
+		gomega.Expect(err).Should(gomega.BeNil(), fmt.Sprintf("%s: %s", failMessage, utils.Err(err)))
+
+		_, err = h.Kubectl("", "delete", "ns", controllerNs)
+		gomega.Expect(err).Should(gomega.BeNil(), utils.Err(err))
+
+	case contour:
+		_, err := h.Kubectl("", "delete", "--ignore-not-found", "-f", controllerYAML)
+		gomega.Expect(err).Should(gomega.BeNil(), fmt.Sprintf("%s: %s", failMessage, utils.Err(err)))
+
+	default:
+		_, err := h.Kubectl("", "delete", "-f", controllerYAML)
+		gomega.Expect(err).Should(gomega.BeNil(), fmt.Sprintf("%s: %s", failMessage, utils.Err(err)))
+	}
+}
+
+func testIngress(ingressName, deploy, ns, resourceYAMLPath string) {
+	h, _ := utils.GetHelperAndConfig()
 
 	ginkgo.By("Applying ingress resource")
-	_, err = h.Kubectl("", "apply", "-f", "testdata/ingress/resources/nginx.yaml")
-	gomega.Expect(err).Should(gomega.BeNil(), fmt.Sprintf("failed to create ingress resource: %s", utils.Err(err)))
+	_, err := h.Kubectl("", "apply", "-f", resourceYAMLPath)
+	gomega.Expect(err).Should(gomega.BeNil(), fmt.Sprintf("failed to create %s ingress resource: %s", ingressName, utils.Err(err)))
 
 	ginkgo.By("Checking if emojivoto is reachable")
-	ip, err := getExternalIP(utils.NginxController, utils.NginxNs)
+	ip, err := getExternalIP(deploy, ns)
```

Comment:
I ran the tests locally on Kind, and my test failed stating 
```
/conformance/specs/specs.go:58

  /conformance/specs/specs.go:38
    ingress
    /conformance/specs/ingress/spec.go:81
      nginx:
      /conformance/specs/ingress/spec.go:89
        nginx ingress controller must work with Linkerd [It]
        /conformance/specs/ingress/spec.go:94

        IP address is empty
        Expected
            <*errors.errorString | 0xc0000fc8d0>: {
                s: "IP address is empty",
            }
        to be nil
```
as there is no controller to create external IPs.

_(+1 on the verbose logs)_

Should we instead use the `clusterIP` to reach out to the service and test? :thinking: but that could not work when we are running the tests directly and not through sonobuoy. So, we try for external IP and fall back to clusterIP if there isn't one? :thinking: 

cc'ing @ihcsim to see if he has any better suggestions here.

Reply:
Hmm, if I understand correctly, the point of ingress is to allow external traffic from the internet into the cluster. `ClusterIP` only exposes a service internally so that it is accessible only from within the cluster, which defeats the point of having an ingress. We would expect users to test ingress only when they know that their cluster supports having a `LoadBalancer` with an external IP as required by Ingress. WDYT?

Reply:
I can definitely imagine a lot of cluster env's (could be on-premise, etc) where they don't get external IPs directly. Ingress is definitely tightly coupled with having a external IP but it also very much used to have a single point of access to all services. My worry is that, we would just miss running Ingress tsts on a lot of environments because of this expectation.

I will leave the decision to @ihcsim :)

Reply:
Using `ClusterIP` is fine, as @Pothulapati suggested. The most important thing is that the `l5d-dst-override` header is added to all requests which go through the ingress controller., regardless of the request origin. This should work because the relevant annotation defined in the `Ingress` resource should get converted into the ingress controller's config.

The ingress-nginx e2e test uses the clusterIP of the nginx controller pod too:

1. Create a HTTP client [here](https://github.com/kubernetes/ingress-nginx/blob/030c15b32194a142e7262db0e97b98ed143f845f/test/e2e/ingress/pathtype_exact.go#L67-L73)
1. Set the base target URL in the client [here](https://github.com/kubernetes/ingress-nginx/blob/030c15b32194a142e7262db0e97b98ed143f845f/test/e2e/framework/framework.go#L458)
1. Which is really just clusterIP of the service as seen [here](https://github.com/kubernetes/ingress-nginx/blob/030c15b32194a142e7262db0e97b98ed143f845f/test/e2e/framework/framework.go#L182-L190)